File: src/tables/avar.cpp
#include "tables.h"

void parseAvar(Parser &parser)
{
    const auto majorVersion = parser.read<UInt16>("Major version");
    const auto minorVersion = parser.read<UInt16>("Minor version");
    if (!(majorVersion == 1 && minorVersion == 0)) {
        throw QString("invalid table version");
    }

    parser.read<UInt16>("Reserved");
    const auto axisCount = parser.read<UInt16>("Axis count");
    parser.readArray("Axes", axisCount, [&](const auto index){
        parser.beginGroup(QString("Segment Map %1").arg(index));
        const auto pairsCount = parser.read<UInt16>("Number of map pairs");
        parser.readArray("Pairs", pairsCount, [&](const auto index2){
            parser.beginGroup(index2);
            parser.read<F2DOT14>("From coordinate");
            parser.read<F2DOT14>("To coordinate");
            parser.endGroup();
        });

        parser.endGroup();
    });
}

File: core/src/tables/avar.rs
use crate::parser::*;
use crate::{Error, Result};

pub fn parse(parser: &mut Parser) -> Result<()> {
    let major_version = parser.read::<u16>("Major version")?;
    let minor_version = parser.read::<u16>("Minor version")?;
    if !(major_version == 1 && minor_version == 0) {
        return Err(Error::InvalidTableVersion);
    }

    parser.read::<u16>("Reserved")?;
    let axis_count = parser.read::<u16>("Axis count")?;
    for _ in 0..axis_count {
        parser.begin_group("Segment map");
        let pairs_count = parser.read::<u16>("Number of map pairs")?;
        for j in 0..pairs_count {
            parser.begin_group(format!("Pair {}", j));
            parser.read::<F2DOT14>("From coordinate")?;
            parser.read::<F2DOT14>("To coordinate")?;
            parser.end_group();
        }

        parser.end_group();
    }

    Ok(())
}
